test/integration/test_workflow_refactoring.py

```diff
@@ -887,3 +895,7 @@ def __init__(self, app, user):
     @property
     def app(self):
         return self._app
+
+    @property
+    def url_builder(self) -> None:
+        return None
```

Comment:
Not having this wasn't an error at runtime because `ProvidesAppContext` doesn't have `metaclass=abc.ABCMeta`, should we add that?

Reply:
I think ProvidesAppContext is meant to be what would now be called a Protocol. I'm not sure I want to take on the project of ensuring the metadata class stuff is okay? Is that okay?